### PR TITLE
resumable send/receive and auto-retry feature

### PR DESCRIPTION
**Pull request by Fmajor**
Hi there,
I am developing my zfs based auto-backup workflow. I have been looking for zfs manage scripts with resumable send/receive and auto-retry feature for some time, but did not find one. So i decide to write these features by myself.
Here is the logs of my test
```
#========== important comments in between==========
#========== config file =================
$ cat /etc/pyznap/pyznap.conf
[ssh:22:root@remote:t]
snap = yes
clean = yes
dest = t/backups/remote/t
resume = yes
dest_auto_create = yes
ServerAliveInterval = 10
ServerAliveCountMax = 1
retry = 10
retry_interval = 10

#========== local status =================
$ zfs list
NAME   USED  AVAIL     REFER  MOUNTPOINT
t      876K  6.38G     48.8K  /t

#========== remote status =================
$ ssh remote zfs list -t all
NAME             USED  AVAIL     REFER  MOUNTPOINT
t                852M  5.55G      150M  /t
t@0M            30.2K      -     46.5K  -
t@50M           30.2K      -     50.1M  -
t@100M          30.2K      -      100M  -
t@150M          32.6K      -      150M  -
t/a              301M  5.55G     46.5K  /t/a
t/a@0M          30.2K      -     46.5K  -
t/a/b            300M  5.55G     46.5K  /t/a/b
t/a/b@0M        30.2K      -     46.5K  -
t/a/b/c          300M  5.55G     46.5K  /t/a/b/c
t/a/b/c@0M      30.2K      -     46.5K  -
t/a/b/c/d        300M  5.55G      300M  /t/a/b/c/d
t/a/b/c/d@0M    32.6K      -     46.5K  -
t/a/b/c/d@50M   32.6K      -     50.1M  -
t/a/b/c/d@100M  30.2K      -      100M  -
t/a/b/c/d@150M  32.6K      -      150M  -
t/a/b/c/d@200M  30.2K      -      200M  -
t/a/b/c/d@250M  30.2K      -      250M  -
t/a/b/c/d@300M     0B      -      300M  -
t/large          400M  5.55G      400M  /t/large
t/large@200M    32.6K      -      200M  -
t/large@400M       0B      -      400M  -

$ pyznap  send
Sep 04 04:19:02 INFO: Starting pyznap...
Sep 04 04:19:02 INFO: Sending snapshots...
Sep 04 04:19:02 WARNING: lzop does not exist, continuing without compression...
#========== new feature: create dest parent folders if non-exists =================
Sep 04 04:19:03 INFO: Destination t/backups/remote/t does not exist...create
Sep 04 04:19:03 INFO: No common snapshots on t/backups/remote/t, sending oldest snapshot root@remote:t@0M (~12.6K)...
Sep 04 04:19:03 INFO: Updating t/backups/remote/t with recent snapshot root@remote:t@150M (~150.3M)...
#========== Ctrl+C to stop, test resume feature =================
^CSep 04 04:19:19 ERROR: KeyboardInterrupt while sending to t/backups/remote/t...
Sep 04 04:19:19 ERROR: KeyboardInterrupt - exiting gracefully...

$ pyznap  send
Sep 04 04:19:21 INFO: Starting pyznap...
Sep 04 04:19:21 INFO: Sending snapshots...
Sep 04 04:19:21 WARNING: lzop does not exist, continuing without compression...
#========== new feature: resume from last transfer =================
Sep 04 04:19:21 INFO: Resume last transfer of t/backups/remote/t (~8.9M), then update again...
Sep 04 04:19:23 INFO: Updating t/backups/remote/t with recent snapshot root@remote:t@150M (~50.1M)...
Sep 04 04:19:31 INFO: t/backups/remote/t is up to date...
Sep 04 04:19:31 INFO: No common snapshots on t/backups/remote/t/a, sending oldest snapshot root@remote:t/a@0M (~12.6K)...
Sep 04 04:19:31 INFO: t/backups/remote/t/a is up to date...
Sep 04 04:19:32 INFO: No common snapshots on t/backups/remote/t/a/b, sending oldest snapshot root@remote:t/a/b@0M (~12.6K)...
Sep 04 04:19:32 INFO: t/backups/remote/t/a/b is up to date...
Sep 04 04:19:32 INFO: No common snapshots on t/backups/remote/t/a/b/c, sending oldest snapshot root@remote:t/a/b/c@0M (~12.6K)...
Sep 04 04:19:32 INFO: t/backups/remote/t/a/b/c is up to date...
Sep 04 04:19:32 INFO: No common snapshots on t/backups/remote/t/a/b/c/d, sending oldest snapshot root@remote:t/a/b/c/d@0M (~12.6K)...
Sep 04 04:19:32 INFO: Updating t/backups/remote/t/a/b/c/d with recent snapshot root@remote:t/a/b/c/d@300M (~300.6M)...
#========== Ctrl+C to stop, test resume feature =================
^CSep 04 04:19:41 ERROR: KeyboardInterrupt while sending to t/backups/remote/t/a/b/c/d...
Sep 04 04:19:41 ERROR: KeyboardInterrupt - exiting gracefully...

$ pyznap  send
Sep 04 04:19:42 INFO: Starting pyznap...
Sep 04 04:19:42 INFO: Sending snapshots...
Sep 04 04:19:42 WARNING: lzop does not exist, continuing without compression...
Sep 04 04:19:43 INFO: t/backups/remote/t is up to date...
Sep 04 04:19:43 INFO: t/backups/remote/t/a is up to date...
Sep 04 04:19:43 INFO: t/backups/remote/t/a/b is up to date...
Sep 04 04:19:43 INFO: t/backups/remote/t/a/b/c is up to date...
#========== new feature: resume from last transfer =================
Sep 04 04:19:43 INFO: Resume last transfer of t/backups/remote/t/a/b/c/d (~48.8M), then update again...
Sep 04 04:19:52 INFO: Updating t/backups/remote/t/a/b/c/d with recent snapshot root@remote:t/a/b/c/d@300M (~200.4M)...
#========== here i shutdown the network  =================
#========== The default ssh config will make ssh freeze almost forever when server is lost =================
#========== I have change the ssh config so that it will exit with error in 10s when server has no response =================
#========== new feature: retry when got ssh connection error =================
Sep 04 04:20:28 WARNING: SSH connection lost, sleep 10 and retry 1/10
Sep 04 04:20:38 WARNING: SSH connection lost, sleep 10 and retry 2/10
Sep 04 04:20:48 WARNING: SSH connection lost, sleep 10 and retry 3/10
#========== here i resume the network  =================
#========== again resume from last transfer  =================
Sep 04 04:20:58 INFO: Resume last transfer of t/backups/remote/t/a/b/c/d (~12.2M), then update again...
Sep 04 04:21:01 INFO: Updating t/backups/remote/t/a/b/c/d with recent snapshot root@remote:t/a/b/c/d@300M (~100.2M)...
Sep 04 04:21:18 INFO: t/backups/remote/t/a/b/c/d is up to date...
Sep 04 04:21:18 INFO: No common snapshots on t/backups/remote/t/large, sending oldest snapshot root@remote:t/large@200M (~200.4M)...
#========== Ctrl+C to stop, test resume feature =================
^CSep 04 04:21:32 ERROR: KeyboardInterrupt while sending to t/backups/remote/t/large...
Sep 04 04:21:32 ERROR: KeyboardInterrupt - exiting gracefully...

#========== last run to finish the send =================
$ pyznap  send
Sep 04 04:21:34 INFO: Starting pyznap...
Sep 04 04:21:34 INFO: Sending snapshots...
Sep 04 04:21:34 WARNING: lzop does not exist, continuing without compression...
Sep 04 04:21:35 INFO: t/backups/remote/t is up to date...
Sep 04 04:21:35 INFO: t/backups/remote/t/a is up to date...
Sep 04 04:21:35 INFO: t/backups/remote/t/a/b is up to date...
Sep 04 04:21:35 INFO: t/backups/remote/t/a/b/c is up to date...
Sep 04 04:21:35 INFO: t/backups/remote/t/a/b/c/d is up to date...
Sep 04 04:21:35 INFO: Resume last transfer of t/backups/remote/t/large (~125.1M)...
Sep 04 04:21:57 INFO: Updating t/backups/remote/t/large with recent snapshot root@remote:t/large@400M (~125.1M)...
Sep 04 04:22:35 INFO: t/backups/remote/t/large is up to date...
Sep 04 04:22:35 INFO: Finished successfully...

#========== done =================
```
Let me explain the changes in the next two threads to help you review my commits

### CODE REVIEW

**Comment by Fmajor** [comment]
# add dest_auto_create and resume feature
I add new options:
* --resume
* --dest-auto-create
currently the resume feature only wolk when dest-auto-create is set to true
## dest_auto_create feature
--dest-auto-create config will automatically create the non-exists parents of dest, 
e.g., receive remote:t to local:t/backups/remote/t, will not raise DatasetNotFoundError, but create its parents instead.
## resume feature
checkout the `receive_resume_token` before send, rerun update for this dataset again when last transfer is resumed.

**Comment by yboetz** [comment]
Hi, thanks a lot for this! I'm on vacation until mid next week, but I'll review it as soon as I'm back.

**Comment by Fmajor** [comment]
# SSH timeout options and retry feature
I add new options:
* --retry
* --retry-interval
I add new options to the config file
* ServerAliveInterval
* ServerAliveCountMax
they are options for `ssh`. 
The default settings of them make `ssh` froze almost forever when the network is lost.
So we need the `ssh` to quit in time and retry the transfer

**Comment by Fmajor** [comment]
The current code is backward compatible, the resume and retry feature will work only when --dest-auto-create is set.
When the code is fully reviewed and tested, I think the --resume feature should be default.

File: pyznap/process.py [hunk 1]
```diff
@@ -65,6 +65,10 @@ def check_returncode(self):
                               HoldTagExistsError):
                     if reason == error.strerror:
                         raise error(dataset)
+        elif self.stderr.startswith('ssh:'):
```
Comment (Fmajor) [inline pyznap/process.py:68]: this should be included in the next commit...

**Comment by yboetz** [comment]
So, I finally started to look into this. Looking good so far, though I will have some questions and changes tomorrow. I'll also have to test it thoroughly, and maybe add a test as well about the resume feature.

File: pyznap/utils.py [hunk 2]
```diff
@@ -100,9 +102,12 @@ def read_config(path):
             else:
                 if option in ['key']:
                     dic[option] = value if os.path.isfile(value) else None
-                elif option in ['frequent', 'hourly', 'daily', 'weekly', 'monthly', 'yearly']:
+                elif option in [
+                    'frequent', 'hourly', 'daily', 'weekly', 'monthly', 'yearly',
+                    'retry', 'retry_interval', 'ServerAliveInterval', 'ServerAliveCountMax'
```
Comment (yboetz) [inline pyznap/utils.py:107]: From my comment in the `main.py` file: If these are set as per-dest values, then this will not work anymore, we will have to treat them similar to the `raw_send` option.

File: pyznap/utils.py [hunk 3]
```diff
@@ -100,9 +102,12 @@ def read_config(path):
...
+                ]:
                     dic[option] = int(value)
-                elif option in ['snap', 'clean']:
+                elif option in ['snap', 'clean', 'resume', 'dest_auto_create']:
```
Comment (yboetz) [inline pyznap/utils.py:110]: Same here.

File: pyznap/send.py [hunk 4]
```diff
@@ -38,19 +39,19 @@ def send_snap(snapshot, dest_name, base=None, ssh_dest=None, raw=False):
     Returns
     -------
     int
-        0 if success, 1 if not
+        0 if success, 1 if not, 2 if resume from last transfer
```
Comment (yboetz) [inline pyznap/send.py:42]: Where does it return 2?

File: pyznap/send.py [hunk 5]
```diff
@@ -64,10 +65,14 @@ def send_snap(snapshot, dest_name, base=None, ssh_dest=None, raw=False):
         send.stderr.close()
 
         stdout, stderr = recv.communicate()
+        if stderr.decode().startswith('cannot receive incremental stream'):
+            raise SSHConnectError(stderr)
```
Comment (yboetz) [inline pyznap/send.py:69]: Why not let this be handled by the next line with `CalledProcessError`?

File: pyznap/send.py [hunk 6]
```diff
@@ -64,10 +65,14 @@ def send_snap(snapshot, dest_name, base=None, ssh_dest=None, raw=False):
...
         # raise any error that occured
         if recv.returncode:
             raise CalledProcessError(returncode=recv.returncode, cmd=recv.args, output=stdout, stderr=stderr)
 
+    except SSHConnectError as err:
+        raise
```
Comment (yboetz) [inline pyznap/send.py:75]: This method should not raise an error it doesn't handle itself, except for the `KeyboardInterrupt`, which is then handled further up. This leads to a crash in pyznap if you do the following for example:

    $ pyznap send -s test0 -d test1 -r
    $ ctrl+C
    $ pyznap send -s test0 -d test1

The second `pyznap send` fails as there is a partial receive token and thus raises the `SSHConnectError`, but this is not handled anywhere in the code, so pyznap crashes.

Comment (yboetz) [reply pyznap/send.py:75]: I think we should remove the `SSHConnectError` and maybe add another exception class for `cannot receive incremental stream`, as said above.

**Review by yboetz — COMMENTED** [review]
Most of it looks good, thanks a lot for the work! I'm not happy yet with the code in `send.py`, there seems to be a bit too much code duplication and not all cases are handled correctly I think. I'll have to go through it in more detail. Can I push changes directly to your `Fmajor:master` branch?

File: pyznap/process.py (hunk 1 above)
Comment (Fmajor) [reply pyznap/process.py:68]: I found that ssh will always return its error message with "ssh:*", it could be a connection error or something else. Maybe we could use a more rubost criterion to test the connection error.
Anyway, when ssh report a error, the connection must be lost, this criterion works in my test, but may be improved.

File: pyznap/process.py [hunk 7]
```diff
@@ -65,6 +65,10 @@ def check_returncode(self):
...
+            # will have circular import problem if put this import at top
+            from .ssh import SSHConnectError
+            raise SSHConnectError(self.stderr)
```
[the remark this answers is not in the document]
Comment (Fmajor) [reply pyznap/process.py:71]: i will get circular import problem if i put `from .ssh import SSHConnectError` at top

File: pyznap/pyzfs.py [hunk 8]
```diff
@@ -347,11 +349,11 @@ def clone(self, name, props={}, force=False):
         raise NotImplementedError()
 
     def send(self, ssh_dest=None, base=None, intermediates=False, replicate=False,
-             properties=False, deduplicate=False, raw=False):
+             properties=False, deduplicate=False, raw=False, receive_resume_token=None):
```
[the remark this answers is not in the document]
Comment (Fmajor) [reply pyznap/pyzfs.py:352]: I think receive_resume_token is better, because it's one `zfs property` if this dataset has a interrupted recieve.

File: pyznap/ssh.py [hunk 9]
```diff
@@ -20,6 +20,10 @@ class SSHException(Exception):
     """General ssh exception to be raised if anything fails"""
     pass
 
+class SSHConnectError(SSHException):
+    """General ssh exception to be raised if anything fails"""
+    pass
+
```
[the remark this answers is not in the document]
Comment (Fmajor) [reply pyznap/ssh.py:26]: Yes, you are right. I did not test all kinds of SSH Exceptions, so we can use  `SSHException` directly.

File: pyznap/ssh.py [hunk 10]
```diff
@@ -79,9 +83,16 @@ def __init__(self, user, host, key=None, port=22, compress=None):
             raise FileNotFoundError(self.key)
 
         self.cmd = ['ssh', '-i', self.key, '-o', 'ControlMaster=auto', '-o', 'ControlPersist=1m',
-                    '-o', 'ControlPath={:s}'.format(self.socket), '-p', str(self.port), 
+                    '-o', 'ControlPath={:s}'.format(self.socket), '-p', str(self.port),
                     '{:s}@{:s}'.format(self.user, self.host)]
 
+        if kwargs.get("ServerAliveInterval"):
+            self.cmd.insert(1, '-o')
+            self.cmd.insert(2, 'ServerAliveInterval={}'.format(kwargs.get('ServerAliveInterval')))
```
[the remark this answers is not in the document]
Comment (Fmajor) [reply pyznap/ssh.py:91]: Oh, new syntax sugar leanred from you ^_^

File: pyznap/utils.py [hunk 11]
```diff
@@ -85,7 +85,9 @@ def read_config(path):
 
     config = []
     options = ['key', 'frequent', 'hourly', 'daily', 'weekly', 'monthly', 'yearly', 'snap', 'clean',
-               'dest', 'dest_keys', 'compress', 'exclude', 'raw_send']
+               'dest', 'dest_keys', 'compress', 'exclude', 'raw_send', 'resume', 'dest_auto_create',
+               'retry', 'retry_interval', 'ServerAliveInterval', 'ServerAliveCountMax',
```
[the remark this answers is not in the document]
Comment (Fmajor) [reply pyznap/utils.py:89]: `retry` is the number of retries.
`retry_interval` is the seconds between two retries
`ServerAliveInterval x ServerAliveCountMax` is the max time ssh wait if server have no response. We may set default value of them. (10,3) may be a good choice. `ssh` will sent heart beats every 10s and if get no response for 3 times, it will exit with error.

The default settings of `ServerAliveInterval` and  `ServerAliveCountMax` in a unconfigured `ssh` will make it freeze almost forever when the network is lost (ssh just wait for server for very long time and raise no error)
So we need the ssh to quit in time and retry the transfer

File: pyznap/send.py (hunk 4 above)
Comment (Fmajor) [reply pyznap/send.py:42]: add this doc in the wrong place...

File: pyznap/send.py (hunk 5 above)
Comment (Fmajor) [reply pyznap/send.py:69]: I need the SSHConnectError catched by `send_config` in line `337`
when we lost connection, we should do a new `send_filesystem`, if here we raise `CalledProcessError`, `send_snap` and `send_filesystem` will return with 1, `send_config` will think we just get error when send this particular dataset, and will skip it without retry.

File: pyznap/send.py [hunk 12]
```diff
@@ -268,10 +327,56 @@ def send_config(config):
...
-                    send_filesystem(source_fs, dest_name, ssh_dest=ssh_dest, raw=raw)
-            finally:
+                    retries = 0
+                    while True: # retry when lose SSH connection
+                        try:
+                            status = send_filesystem(source_fs, dest_name, ssh_dest=ssh_dest, raw=raw, resume=resume)
+                            if status==2: # just resume from last transfer, do upate again
+                                send_filesystem(source_fs, dest_name, ssh_dest=ssh_dest, raw=raw, resume=resume)
+                            break
+                        except SSHConnectError as err:
```
Comment (Fmajor) [inline pyznap/send.py:337]: This is the key point for the retry

**Review by Fmajor — COMMENTED** [review]
> Most of it looks good, thanks a lot for the work! I'm not happy yet with the code in `send.py`, there seems to be a bit too much code duplication and not all cases are handled correctly I think. I'll have to go through it in more detail. Can I push changes directly to your `Fmajor:master` branch?

Sure, I'd like my code to follow your style. I just finished a very rough framework for these functions, you are the one to decide the final version.

File: pyznap/send.py (hunk 6 above)
Comment (Fmajor) [reply pyznap/send.py:75]: > This method should not raise an error it doesn't handle itself, except for the `KeyboardInterrupt`, which is then handled further up. This leads to a crash in pyznap if you do the following for example:
> 
> ```
> $ pyznap send -s test0 -d test1 -r
> $ ctrl+C
> $ pyznap send -s test0 -d test1
> ```
> 
> The second `pyznap send` fails as there is a partial receive token and thus raises the `SSHConnectError`, but this is not handled anywhere in the code, so pyznap crashes.

can you fix this bug?

Comment (yboetz) [reply pyznap/send.py:75]: Yes, will do.

**Comment by yboetz** [comment]
> Sure, I'd like my code to follow your style. I just finished a very rough framework for these functions, you are the one to decide the final version.

Great. I'll work on this over the next few days and try to keep you up to date.

File: pyznap/send.py (hunk 6 above)
Comment (Fmajor) [reply pyznap/send.py:75]: **One important thing**
`74 except SSHConnectError as err:`
this exception is either raised by `send`(in line 68, process.py) or `recv`(in line 68, send.py), I don't know which one will raise the error first, so i catch both.
`send` will raise an 'SSHConnectError' and `recv` will have error msg: `cannot receive incremental stream`
That's the reason I choose `SSHConnectError` as the name here, because `send`(in line 68, process.py) will raise a `SSHConnectError`

**Comment by yboetz** [comment]
I've made some changes, though not finished yet. Could you give me permission to push to this branch?

**Comment by yboetz** [comment]
It works pretty well now, seems to handle the corner cases I could think of. I still need to update the sample config and documentation, and maybe add a test... Hopefully we can get this merged soon though, I'm excited on how well it works :). Thanks a lot.

File: pyznap/send.py [hunk 13]
```diff
@@ -265,114 +262,134 @@ def send_config(config):
...
+                    logger.error('Send was not successful...')
+
+            if ssh_dest:
+                ssh_dest.close()
 
         if ssh_source:
             ssh_source.close()
+
+
+def create_dataset(name, name_log, ssh=None):
```
Comment (Fmajor) [inline pyznap/send.py:328]: one small bug if you use `zfs.create(name, force=True)` here
if you are non-root, `zfs create -p` will exit 1 and errmsg `filesystem successfully created, but it may only be mounted by root`, even when you have `mount` permission
```
random@gentoo$ zfs allow t                                                                                                                                                                                                                                      
---- Permissions on t ------------------------------------------------
Local+Descendent permissions:
        user random create,destroy,hold,mount,readonly,receive,rename,rollback,send,snapshot,userprop
random@gentoo$ zfs create -p t/backups/remote/new # this command exit with 1                                                                                                                                                                                      
filesystem successfully created, but it may only be mounted by root
random@gentoo$ zfs create -p t/backups/remote/new # this command exit with 0
```
that will stop the `send` if you do not have the parent dataset of your dest, see the test below
```
# here, dest is `t/backups/remote/t`
random@gentoo ~/github/pyznap $ zfs destroy -r t/backups
random@gentoo ~/github/pyznap $ zfs list
NAME              USED  AVAIL     REFER  MOUNTPOINT
t                 947K  6.38G     46.5K  /t
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:06:52 INFO: Starting pyznap...
Sep 19 10:06:52 INFO: Sending snapshots...
Sep 19 10:06:52 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:06:52 INFO: Destination t/backups/remote/t does not exist, will create it...
Sep 19 10:06:53 ERROR: Error while creating t/backups/remote/t: Command '['zfs', 'create', '-p', 't/backups/remote/t']' returned non-zero exit status 1....
Sep 19 10:06:53 INFO: Finished successfully...

random@gentoo ~/github/pyznap $ zfs list
NAME              USED  AVAIL     REFER  MOUNTPOINT
t                1007K  6.38G     46.5K  /t
t/backups        46.5K  6.38G     46.5K  /t/backups
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:06:58 INFO: Starting pyznap...
Sep 19 10:06:58 INFO: Sending snapshots...
Sep 19 10:06:58 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:06:58 INFO: Destination t/backups/remote/t does not exist, will create it...
Sep 19 10:06:58 ERROR: Error while creating t/backups/remote/t: Command '['zfs', 'create', '-p', 't/backups/remote/t']' returned non-zero exit status 1....
Sep 19 10:06:58 INFO: Finished successfully...

random@gentoo ~/github/pyznap $ zfs list
NAME               USED  AVAIL     REFER  MOUNTPOINT
t                 1021K  6.38G     48.8K  /t
t/backups         93.0K  6.38G     46.5K  /t/backups
t/backups/remote  46.5K  6.38G     46.5K  /t/backups/remote
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:07:03 INFO: Starting pyznap...
Sep 19 10:07:03 INFO: Sending snapshots...
Sep 19 10:07:03 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:07:03 INFO: Destination t/backups/remote/t does not exist, will create it...
Sep 19 10:07:03 ERROR: Error while creating t/backups/remote/t: Command '['zfs', 'create', '-p', 't/backups/remote/t']' returned non-zero exit status 1....
Sep 19 10:07:03 INFO: Finished successfully...

random@gentoo ~/github/pyznap $ zfs list
NAME                 USED  AVAIL     REFER  MOUNTPOINT
t                   1.06M  6.38G     51.2K  /t
t/backups            140K  6.38G     46.5K  /t/backups
t/backups/remote    93.0K  6.38G     46.5K  /t/backups/remote
t/backups/remote/t  46.5K  6.38G     46.5K  /t/backups/remote/t
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:07:13 INFO: Starting pyznap...
Sep 19 10:07:13 INFO: Sending snapshots...
Sep 19 10:07:13 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:07:13 INFO: No common snapshots on t/backups/remote/t, sending oldest snapshot root@remote:t@0M (~12.6K)...
Sep 19 10:07:14 INFO: Updating t/backups/remote/t with recent snapshot root@remote:t@150M (~150.3M)...
^CSep 19 10:07:15 ERROR: KeyboardInterrupt while sending to t/backups/remote/t...
Sep 19 10:07:15 ERROR: KeyboardInterrupt - exiting gracefully...
```

my origin code about creating ancestors datasets is ugly, but work with this situation: ancestors of dest dataset not existing and you are non-root

**Comment by Fmajor** [comment]
**retry not work properly**
```
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:22:29 INFO: Starting pyznap...
Sep 19 10:22:29 INFO: Sending snapshots...
Sep 19 10:22:30 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:22:30 INFO: Found resume token. Resuming last transfer of t/backups/remote/t (~33.4M)...
#=======================here cut off network========================
Sep 19 10:23:13 ERROR: Error while sending to t/backups/remote/t: cannot receive resume stream: checksum mismatch or incomplete stream. - Partially received snapshot is saved. - A resuming stream can be generated on the sending system by running: -     zfs send -t 1-ef234f49d-d8-789c636064000310a501c49c50360710a715e5e7a69766a6304041ec5ba9fed347323e2900d9ec48eaf293b252934b800c16883a0cf9b4b4e254903c4309134c9e0d493ea9b224b518484bacaa64c2a6bf241fe20a2315369e2959ef5b90e539c1f27989b9a940dac1d0c0c017e2060017cc1bbd...
Sep 19 10:23:13 INFO: Retrying send in 10s (retry 1 of 10)...
Traceback (most recent call last):
  File "/home/random/.pyenv/versions/3.7.3/bin/pyznap", line 11, in <module>
    load_entry_point('pyznap', 'console_scripts', 'pyznap')()
  File "/home/random/github/pyznap/pyznap/main.py", line 170, in main
    return _main()
  File "/home/random/github/pyznap/pyznap/main.py", line 153, in _main
    send_config(config)
  File "/home/random/github/pyznap/pyznap/send.py", line 314, in send_config
    rc = send_filesystem(source_fs, dest_name, ssh_dest=ssh_dest, raw=raw, resume=resume)
  File "/home/random/github/pyznap/pyznap/send.py", line 118, in send_filesystem
    snapshots = source_fs.snapshots()[::-1]
  File "/home/random/github/pyznap/pyznap/pyzfs.py", line 222, in snapshots
    return find(self.name, ssh=self.ssh, max_depth=1, types=['snapshot'])
  File "/home/random/github/pyznap/pyznap/pyzfs.py", line 59, in find
    out = check_output(cmd, ssh=ssh)
  File "/home/random/github/pyznap/pyznap/process.py", line 105, in check_output
    ret.check_returncode()
  File "/home/random/github/pyznap/pyznap/process.py", line 70, in check_returncode
    super(CompletedProcess, self).check_returncode()
  File "/home/random/.pyenv/versions/3.7.3/lib/python3.7/subprocess.py", line 428, in check_returncode
    self.stderr)
subprocess.CalledProcessError: Command '['ssh', '-i', '/home/random/.ssh/id_rsa', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=1m', '-o', 'ControlPath=/tmp/pyznap_root@remote:22_2020-09-19_10:22:29', '-p', '22', '-o', 'ServerAliveInterval=10', 'root@remote', 'zfs', 'list', '-H', '-d', '1', '-t', 'snapshot', '-o', 'name,type', 't']' returned non-zero exit status 255.
```

File: pyznap/ssh.py [hunk 14]
```diff
@@ -84,14 +84,7 @@ def __init__(self, user, host, key=None, port=22, compress=None, **kwargs):
...
                     '-o', 'ControlPath={:s}'.format(self.socket), '-p', str(self.port),
-                    '{:s}@{:s}'.format(self.user, self.host)]
-
-        if kwargs.get("ServerAliveInterval"):
-            self.cmd.insert(1, '-o')
-            self.cmd.insert(2, 'ServerAliveInterval={}'.format(kwargs.get('ServerAliveInterval')))
-        if kwargs.get("ServerAliveCountMax"):
-            self.cmd.insert(1, '-o')
-            self.cmd.insert(2, 'ServerAliveCountMax={}'.format(kwargs.get('ServerAliveCountMax')))
+                    '-o', 'ServerAliveInterval=60', '{:s}@{:s}'.format(self.user, self.host)]
```
Comment (Fmajor) [inline pyznap/ssh.py:87]: can we add the `ServerAliveInterval` and `ServerAliveCountMax` config back some where?
I think 60s is too long sometimes (e.g., when debuging the retry feature, i want 10s)

File: pyznap/send.py (hunk 13 above)
Comment (yboetz) [reply pyznap/send.py:328]: Hm yes, I didn't consider this, since I always work with root access. This seems like it is a zfs bug (see e.g. https://github.com/openzfs/zfs/issues/8478), since I cannot even create datasets with `zfs create -p -o mountpoint=none test/a/b`, I always get permission problems. I'd rather not have too much code in pyznap handling these special cases, since most of the time you should just create the dataset manually instead of letting pyznap without root access handle it. But what I did now:

```
    try:
        zfs.create(name, ssh=ssh, force=True)
    except CalledProcessError as err:
        message = err.stderr.rstrip()
        if message == "filesystem successfully created, but it may only be mounted by root":
            logger.info('Successfully created {:s}, but cannot mount as non-root...'.format(name_log))
            return 0
        else:
            logger.info('Error while creating {}: \'{:s}\'...'.format(name_log, message))
            return 1
```

So it checks for this error message `filesystem successfully created, but it may only be mounted by root` and will continue with the send if the filesystem was created but not mounted. This means that it can handle creating datasets when the parents exist, but still will fail when the parent not exist. In this case you will have to manually create it.

I also improved the error handling, so it actually shows the error now...

**Comment by yboetz** [comment]
> **retry not work properly**

Thanks, that was a `CalledProcessError` that was not catched when looking for snapshots over ssh if the network is down :). Should be fixed now and correctly abort the send. 

> also, we should make this message to DEBUG level (because the retry feature handle this error properly, An error in log must be a real error)

There is still an error happening, and it would mean more code to have this as DEBUG level if we have retries and as ERROR level if we dont have retries :D.

File: pyznap/ssh.py (hunk 14 above)
Comment (yboetz) [reply pyznap/ssh.py]: Atm I feel strongly towards not adding too many options. I've changed the default to a more reasonable 30s, and for debugging you will have the code open anyway and can quickly set it to something small.

File: pyznap/send.py [hunk 15]
```diff
@@ -316,7 +316,7 @@ def send_config(config):
                 # send not excluded filesystems
                 for retry in range(1,retries+2):
                     rc = send_filesystem(source_fs, dest_name, ssh_dest=ssh_dest, raw=raw, resume=resume)
-                    if rc == 2 and retry <= retries:
+                    if rc != 0 and retry <= retries:
```
Comment (Fmajor) [inline pyznap/send.py:319]: We will only get rc==2 in the first fail of network. In the first retry, we will get other others errors(and the loop return 1). The code will not reach to the `send` and `receive` command.
so just return 1 when we got a error, continue the retry loop when we get the error
```
# test before this change
pyznap$ pyznap send
Sep 19 21:45:15 INFO: Starting pyznap...
Sep 19 21:45:15 INFO: Sending snapshots...
Sep 19 21:45:15 WARNING: lzop does not exist, continuing without compression...
.....
Sep 19 21:45:16 INFO: Found resume token. Resuming last transfer of t/backups/remote/t/large (~154.1M)...
Sep 19 21:45:39 ERROR: Error while sending to t/backups/remote/t/large: cannot receive resume stream: checksum mismatch or incomplete stream. - Partially received snapshot is saved. - A resuming stream can be generated on the sending system by running: -     zfs send -t 1-b9364b1bd-b8-789c636064000310a500c4ec50360710e72765a52697303030419460caa7a515a7968064c45860f26c48f2499525a9c5208989322cd8f497e4a79766a63030cc9428547912d7b94f03499e132c9f97989b0aa4f573128bd2531d8c0c0c7ca14e610000de681599...
Sep 19 21:45:39 INFO: Retrying send in 10s (retry 1 of 10)...
Sep 19 21:45:49 ERROR: Error while opening source root@remote:t/large: 'ssh: Could not resolve hostname 8027.node.office: Name or service not known'...
# no more tries because the code return 1 before it reach `send` and `receive` and return 2
Sep 19 21:45:49 INFO: Finished successfully...
```

**Comment by Fmajor** [comment]
I think the code is good enough. Maybe we can call other people to do more tests before we merge it?

File: pyznap/send.py (hunk 15 above)
Comment (yboetz) [reply pyznap/send.py:319]: I specifically set it to `rc == 2` because I don't want to trigger retries for all errors. E.g. when there are no snapshots on the source the method `send_filesystem` will return 1, but we don't want to retry on that. I'll revert the check to `rc == 2` but will check for ssh errors especially as you suggested in the beginning.

**Comment by yboetz** [comment]
Can you try your ssh send test again now? I think it should work as intended now, without retrying on things like missing snapshots and so on.

**Comment by Fmajor** [comment]
> Can you try your ssh send test again now? I think it should work as intended now, without retrying on things like missing snapshots and so on.

It works!
Now we get a perfect backup system

**Comment by yboetz** [comment]
Ok, this looks good for me now. If you agree as well, I will merge it and release version 1.6.0 of pyznap!

**Comment by Fmajor** [comment]
> Ok, this looks good for me now. If you agree as well, I will merge it and release version 1.6.0 of pyznap!

I ask people in relevant issues to test this patch, maybe we can wait several days for their response and then do the merge.

**Comment by yboetz** [comment]
Merged! I'll prepare release v1.6.0.

**Comment by yboetz** [comment]
Thanks for getting this started, it's fantastic to finally have resumable send/receive!
